File: tools.py
"""
Backend tools for the Autonomous Customer Resolution Agent.
Simulates enterprise APIs: Customer DB, Order API, Inventory API, Policy Rules Engine, and Verification Endpoint.
"""

import json
import os
from datetime import datetime, timezone

BASE_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(BASE_DIR, "data")

CUSTOMERS_FILE = os.path.join(DATA_DIR, "customers.json")
ORDERS_FILE = os.path.join(DATA_DIR, "orders.json")
INVENTORY_FILE = os.path.join(DATA_DIR, "inventory.json")
POLICY_FILE = os.path.join(DATA_DIR, "policy.json")

# Backup store for resetting state
_SEED_CACHE = {}


def _ensure_seed_cache():
    """Cache initial state for reset_world capability."""
    global _SEED_CACHE
    if not _SEED_CACHE:
        _SEED_CACHE = {
            "customers": _read_file(CUSTOMERS_FILE),
            "orders": _read_file(ORDERS_FILE),
            "inventory": _read_file(INVENTORY_FILE),
            "policy": _read_file(POLICY_FILE),
        }


def _read_file(file_path: str):
    with open(file_path, "r", encoding="utf-8") as f:
        return json.load(f)


def _write_file(file_path: str, data):
    with open(file_path, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2)


def reset_world() -> dict:
    """Reset data initial state."""
    _ensure_seed_cache()
    _write_file(CUSTOMERS_FILE, _SEED_CACHE["customers"])
    _write_file(ORDERS_FILE, _SEED_CACHE["orders"])
    _write_file(INVENTORY_FILE, _SEED_CACHE["inventory"])
    _write_file(POLICY_FILE, _SEED_CACHE["policy"])
    return {
        "success": True,
        "message": "World data successfully reset to initial seed state.",
    }


def get_customer(customer_id: str) -> dict:
    """Get customer profile details."""
    customers = _read_file(CUSTOMERS_FILE)
    for cust in customers:
        if cust["customer_id"] == customer_id:
            return {"success": True, "customer": cust}
    return {
        "success": False,
        "error": f"Customer ID '{customer_id}' not found.",
    }


def get_order(order_id: str) -> dict:
    """Get order by ID."""
    orders = _read_file(ORDERS_FILE)
    for order in orders:
        if order["order_id"] == order_id:
            return {"success": True, "order": order}
    return {
        "success": False,
        "error": f"Order ID '{order_id}' not found.",
    }


def get_customer_orders(customer_id: str) -> dict:
    """Get all customer orders."""
    orders = _read_file(ORDERS_FILE)
    cust_orders = [o for o in orders if o["customer_id"] == customer_id]
    return {"success": True, "customer_id": customer_id, "orders": cust_orders}


def check_inventory(sku: str) -> dict:
    """Check item inventory stock."""
    inventory = _read_file(INVENTORY_FILE)
    for item in inventory:
        if item["sku"] == sku:
            in_stock = item["quantity_in_stock"] > 0
            return {
                "success": True,
                "sku": sku,
                "item_name": item["item_name"],
                "quantity_in_stock": item["quantity_in_stock"],
                "in_stock": in_stock,
                "restock_date": item.get("restock_date"),
            }
    return {
        "success": False,
        "error": f"SKU '{sku}' not found in inventory.",
    }


def check_policy(order_id: str, action: str) -> dict:
    """Check policy for action."""
    order_res = get_order(order_id)
    if not order_res["success"]:
        return order_res

    order = order_res["order"]
    policy = _read_file(POLICY_FILE)
    action_upper = action.strip().upper()

    now = datetime.now(timezone.utc)

    # Calculate days since delivery if delivered
    days_since_delivery = None
    if order.get("delivery_date"):
        deliv_dt = datetime.fromisoformat(order["delivery_date"].replace("Z", "+00:00"))
        days_since_delivery = (now - deliv_dt).days

    if action_upper == "REFUND":
        return_window = policy["return_policy"]["return_window_days"]
        max_refund = policy["refund_policy"]["max_auto_refund_amount"]

        if days_since_delivery is not None and days_since_delivery > return_window:
            return {
                "success": True,
                "eligible": False,
                "reason": f"Order delivery date was {days_since_delivery} days ago, exceeding the {return_window}-day return window.",
            }

        if order["total_amount"] > max_refund:
            return {
                "success": True,
                "eligible": False,
                "reason": f"Order amount (${order['total_amount']:.2f}) exceeds max auto-refund limit of ${max_refund:.2f}.",
            }

        return {
            "success": True,
            "eligible": True,
            "reason": "Refund action is permitted under company policy.",
        }

    elif action_upper == "REPLACEMENT":
        return_window = policy["return_policy"]["return_window_days"]

        if days_since_delivery is not None and days_since_delivery > return_window:
            return {
                "success": True,
                "eligible": False,
                "reason": f"Order delivery date was {days_since_delivery} days ago, exceeding the {return_window}-day replacement window.",
            }

        # Check item inventory for order items
        skus_checked = []
        for item in order.get("items", []):
            inv = check_inventory(item["sku"])
            skus_checked.append(inv)
            if not inv.get("in_stock", False):
                return {
                    "success": True,
                    "eligible": False,
                    "reason": f"Replacement item SKU '{item['sku']}' is currently out of stock.",
                    "inventory_details": inv,
                }

        return {
            "success": True,
            "eligible": True,
            "reason": "Replacement action is permitted and stock is available.",
        }

    elif action_upper == "CANCEL":
        allowed = policy["cancellation_policy"]["allowed_statuses"]
        if order["status"] in allowed:
            return {
                "success": True,
                "eligible": True,
                "reason": f"Order status '{order['status']}' is eligible for cancellation.",
            }
        else:
            return {
                "success": True,
                "eligible": False,
                "reason": f"Order status '{order['status']}' cannot be cancelled. Cancellation is only allowed for statuses: {allowed}.",
            }

    return {
        "success": False,
        "error": f"Unknown policy action '{action}'. Valid actions are REFUND, REPLACEMENT, CANCEL.",
    }


def process_refund(order_id: str, amount: float = None) -> dict:
    """Process an order refund."""
    orders = _read_file(ORDERS_FILE)
    target_index = -1
    target_order = None

    for i, order in enumerate(orders):
        if order["order_id"] == order_id:
            target_index = i
            target_order = order
            break

    if target_order is None:
        return {"success": False, "error": f"Order ID '{order_id}' not found."}

    refund_amount = amount if amount is not None else target_order["total_amount"]

    if target_order["status"] == "REFUNDED":
        return {
            "success": False,
            "error": f"Order '{order_id}' has already been refunded.",
        }

    # Execute refund
    timestamp = datetime.now(timezone.utc).isoformat()
    target_order["status"] = "REFUNDED"
    target_order["resolution_state"] = "REFUNDED"
    target_order["refund_amount"] = refund_amount
    target_order["history"].append({
        "timestamp": timestamp,
        "action": "PROCESS_REFUND",
        "details": f"Processed refund of ${refund_amount:.2f}",
    })

    orders[target_index] = target_order
    _write_file(ORDERS_FILE, orders)

    return {
        "success": True,
        "order_id": order_id,
        "refund_amount": refund_amount,
        "status": "REFUNDED",
        "message": f"Refund of ${refund_amount:.2f} successfully processed for order {order_id}.",
    }


def process_replacement(order_id: str, sku: str) -> dict:
    """Process item replacement request."""
    orders = _read_file(ORDERS_FILE)
    target_index = -1
    target_order = None

    for i, order in enumerate(orders):
        if order["order_id"] == order_id:
            target_index = i
            target_order = order
            break

    if target_order is None:
        return {"success": False, "error": f"Order ID '{order_id}' not found."}

    # Check inventory stock for requested SKU
    inv = check_inventory(sku)
    if not inv["success"]:
        return inv

    if not inv["in_stock"]:
        # Planted failure case triggered!
        return {
            "success": False,
            "error": f"Replacement failed: SKU '{sku}' ({inv['item_name']}) is out of stock (quantity: 0).",
            "out_of_stock_sku": sku,
            "restock_date": inv.get("restock_date"),
        }

    # Deduct 1 item from inventory
    inventory = _read_file(INVENTORY_FILE)
    for item in inventory:
        if item["sku"] == sku:
            item["quantity_in_stock"] -= 1
            break
    _write_file(INVENTORY_FILE, inventory)

    # Execute replacement on order
    timestamp = datetime.now(timezone.utc).isoformat()
    target_order["status"] = "REPLACED"
    target_order["resolution_state"] = "REPLACED"
    target_order["replacement_sku"] = sku
    target_order["history"].append({
        "timestamp": timestamp,
        "action": "PROCESS_REPLACEMENT",
        "details": f"Replacement unit ordered for SKU '{sku}'",
    })

    orders[target_index] = target_order
    _write_file(ORDERS_FILE, orders)

    return {
        "success": True,
        "order_id": order_id,
        "replacement_sku": sku,
        "status": "REPLACED",
        "message": f"Replacement for SKU '{sku}' successfully created for order {order_id}.",
    }


def cancel_order(order_id: str) -> dict:
    """Cancel an active order."""
    orders = _read_file(ORDERS_FILE)
    target_index = -1
    target_order = None

    for i, order in enumerate(orders):
        if order["order_id"] == order_id:
            target_index = i
            target_order = order
            break

    if target_order is None:
        return {"success": False, "error": f"Order ID '{order_id}' not found."}

    # Check policy for cancellation
    pol_check = check_policy(order_id, "CANCEL")
    if not pol_check.get("eligible", False):
        return {
            "success": False,
            "error": f"Cancellation blocked by policy: {pol_check.get('reason')}",
        }

    # Execute cancellation
    timestamp = datetime.now(timezone.utc).isoformat()
    target_order["status"] = "CANCELLED"
    target_order["resolution_state"] = "CANCELLED"
    target_order["history"].append({
        "timestamp": timestamp,
        "action": "CANCEL_ORDER",
        "details": "Order cancelled successfully prior to shipping.",
    })

    orders[target_index] = target_order
    _write_file(ORDERS_FILE, orders)

    return {
        "success": True,
        "order_id": order_id,
        "status": "CANCELLED",
        "message": f"Order {order_id} has been successfully cancelled.",
    }


def verify_state(order_id: str) -> dict:
    """Verify current order status."""
    order_res = get_order(order_id)
    if not order_res["success"]:
        return order_res

    order = order_res["order"]
    is_resolved = order["resolution_state"] in ["REFUNDED", "REPLACED", "CANCELLED"]

    return {
        "success": True,
        "order_id": order_id,
        "current_status": order["status"],
        "resolution_state": order["resolution_state"],
        "is_resolved": is_resolved,
        "items": order["items"],
        "total_amount": order["total_amount"],
        "history": order["history"],
    }


# Ensure seed cache is populated on module load
_ensure_seed_cache()

File: src/lib/tools.ts
// ============================================================
// Simulated Enterprise Backend Tools Engine
// TypeScript implementation of the enterprise simulation engine
// Faithful to AWERA-CODE/Agentic-Ai with full state verification
// ============================================================

import { customersData } from '@/lib/data/customers';
import { ordersData } from '@/lib/data/orders';
import { inventoryData } from '@/lib/data/inventory';
import { policiesData } from '@/lib/data/policies';
import {
  Customer,
  Order,
  InventoryItem,
  ToolResult,
  OrderStatus,
} from '@/lib/types';

// ---- In-Memory State (reset-able for deterministic demos) ----
// Deep-clone seed data so mutations don't corrupt the initial dataset

let customers: Customer[] = structuredClone(customersData);
let orders: Order[] = structuredClone(ordersData);
let inventory: InventoryItem[] = structuredClone(inventoryData);

function now(): string {
  return new Date().toISOString();
}

// ---- Reset Tool ----

export function resetWorld(): ToolResult {
  customers = structuredClone(customersData);
  orders = structuredClone(ordersData);
  inventory = structuredClone(inventoryData);
  return {
    success: true,
    message: 'World data successfully reset to initial seed state.',
  };
}

// ---- Customer Tools ----

export function getCustomer(customerId: string): ToolResult {
  const cust = customers.find((c) => c.customer_id === customerId);
  if (!cust) {
    return { success: false, error: `Customer ID '${customerId}' not found.` };
  }
  return { success: true, customer: { ...cust } };
}

export function getCustomerOrders(customerId: string): ToolResult {
  const custOrders = orders.filter((o) => o.customer_id === customerId);
  return { success: true, customer_id: customerId, orders: custOrders.map((o) => ({ ...o })) };
}

// ---- Order Tools ----

export function getOrder(orderId: string): ToolResult {
  const order = orders.find((o) => o.order_id === orderId);
  if (!order) {
    return { success: false, error: `Order ID '${orderId}' not found.` };
  }
  return {
    success: true,
    order: {
      ...order,
      items: order.items.map((i) => ({ ...i })),
      history: order.history.map((h) => ({ ...h })),
    },
  };
}

// ---- Inventory Tools ----

export function checkInventory(sku: string): ToolResult {
  const item = inventory.find((i) => i.sku === sku);
  if (!item) {
    return { success: false, error: `SKU '${sku}' not found in inventory.` };
  }
  const inStock = item.quantity_in_stock > 0;
  return {
    success: true,
    item: { ...item },
    sku: item.sku,
    item_name: item.item_name,
    category: item.category,
    unit_price: item.unit_price,
    in_stock: inStock,
    quantity_available: item.quantity_in_stock,
    quantity_in_stock: item.quantity_in_stock,
    restock_date: item.restock_date,
  };
}

// ---- Policy Tools ----

export function checkPolicy(orderId: string, action: string): ToolResult {
  const order = orders.find((o) => o.order_id === orderId);
  if (!order) {
    return { success: false, error: `Order ID '${orderId}' not found.` };
  }

  const normalizedAction = action.toUpperCase();
  const policies = policiesData;

  // Calculate days since delivery if delivered
  let daysSinceDelivery: number | null = null;
  if (order.delivery_date) {
    const delivDt = new Date(order.delivery_date);
    const msDiff = new Date().getTime() - delivDt.getTime();
    daysSinceDelivery = Math.floor(msDiff / (1000 * 60 * 60 * 24));
  }

  if (normalizedAction === 'REFUND') {
    const returnWindow = policies.return_policy.return_window_days;
    const maxRefund = 500.0;

    if (order.resolution_state !== 'NONE') {
      return {
        success: false,
        allowed: false,
        eligible: false,
        error: `Order '${orderId}' already has resolution state '${order.resolution_state}'. Cannot process refund.`,
        reason: `Order already resolved as ${order.resolution_state}.`,
      };
    }

    if (daysSinceDelivery !== null && daysSinceDelivery > returnWindow) {
      return {
        success: false,
        allowed: false,
        eligible: false,
        error: `Return window of ${returnWindow} days has expired. Order was delivered ${daysSinceDelivery} days ago.`,
        reason: `Delivery date was ${daysSinceDelivery} days ago, exceeding the ${returnWindow}-day window.`,
      };
    }

    if (order.total_amount > maxRefund) {
      return {
        success: false,
        allowed: false,
        eligible: false,
        error: `Order amount ($${order.total_amount.toFixed(2)}) exceeds max auto-refund limit of $${maxRefund.toFixed(2)}.`,
        reason: `Amount exceeds $${maxRefund.toFixed(2)} limit.`,
      };
    }

    // Customer tier VIP bonus check
    const cust = customers.find((c) => c.customer_id === order.customer_id);
    const vipBonus = cust?.tier === 'VIP' ? policies.refund_policy.vip_bonus_percent : 0;

    return {
      success: true,
      allowed: true,
      eligible: true,
      action: 'REFUND',
      order_id: orderId,
      refund_amount: order.total_amount,
      vip_bonus_percent: vipBonus,
      vip_bonus_amount: (order.total_amount * vipBonus) / 100,
      reason: 'Refund action is permitted under company policy.',
      policy: policies.refund_policy,
    };
  }

  if (normalizedAction === 'REPLACEMENT' || normalizedAction === 'REPLACE') {
    const returnWindow = policies.return_policy.return_window_days;

    if (order.resolution_state !== 'NONE') {
      return {
        success: false,
        allowed: false,
        eligible: false,
        error: `Order '${orderId}' already has resolution state '${order.resolution_state}'. Cannot process replacement.`,
        reason: `Order already resolved as ${order.resolution_state}.`,
      };
    }

    if (!policies.replacement_policy.eligible_statuses.includes(order.status)) {
      return {
        success: false,
        allowed: false,
        eligible: false,
        error: `Replacement not allowed for order status '${order.status}'. Eligible statuses: ${policies.replacement_policy.eligible_statuses.join(', ')}.`,
        reason: `Order status '${order.status}' is not eligible for replacement.`,
      };
    }

    if (daysSinceDelivery !== null && daysSinceDelivery > returnWindow) {
      return {
        success: false,
        allowed: false,
        eligible: false,
        error: `Return window of ${returnWindow} days has expired. Order was delivered ${daysSinceDelivery} days ago.`,
        reason: `Delivery date was ${daysSinceDelivery} days ago, exceeding the ${returnWindow}-day window.`,
      };
    }

    return {
      success: true,
      allowed: true,
      eligible: true,
      action: 'REPLACEMENT',
      order_id: orderId,
      requires_inventory_check: true,
      reason: 'Replacement action is permitted within the return window. Inventory availability check required.',
      policy: policies.replacement_policy,
    };
  }

  if (normalizedAction === 'CANCEL' || normalizedAction === 'CANCELLATION') {
    const allowed = policies.cancellation_policy.allowed_statuses;

    if (order.resolution_state !== 'NONE') {
      return {
        success: false,
        allowed: false,
        eligible: false,
        error: `Order '${orderId}' already has resolution state '${order.resolution_state}'.`,
        reason: `Order already resolved as ${order.resolution_state}.`,
      };
    }

    if (allowed.includes(order.status)) {
      return {
        success: true,
        allowed: true,
        eligible: true,
        action: 'CANCEL',
        order_id: orderId,
        refund_amount: order.total_amount,
        reason: `Order status '${order.status}' is eligible for cancellation.`,
        policy: policies.cancellation_policy,
      };
    }

    return {
      success: false,
      allowed: false,
      eligible: false,
      error: `Cancellation not allowed for order status '${order.status}'. Orders can only be cancelled before they enter SHIPPED or DELIVERED status. Allowed statuses: ${allowed.join(', ')}.`,
      reason: `Order status '${order.status}' cannot be cancelled. Only allowed for: ${allowed.join(', ')}.`,
      policy: policies.cancellation_policy,
      order_status: order.status,
    };
  }

  return {
    success: false,
    allowed: false,
    eligible: false,
    error: `Unknown policy action '${action}'. Valid actions are REFUND, REPLACEMENT, CANCEL.`,
  };
}

// ---- Action Tools ----

export function processRefund(orderId: string, amount?: number): ToolResult {
  const orderIdx = orders.findIndex((o) => o.order_id === orderId);
  if (orderIdx === -1) {
    return { success: false, error: `Order ID '${orderId}' not found.` };
  }
  const order = orders[orderIdx];

  // Verify policy check
  const policyCheck = checkPolicy(orderId, 'REFUND');
  if (!policyCheck.success) {
    return policyCheck;
  }

  const refundAmount = amount !== undefined ? amount : order.total_amount;
  if (refundAmount <= 0 || refundAmount > order.total_amount) {
    return {
      success: false,
      error: `Invalid refund amount $${refundAmount.toFixed(2)}. Order total is $${order.total_amount.toFixed(2)}.`,
    };
  }

  // Prevent duplicate refund
  if (order.resolution_state === 'REFUND_PROCESSED' || order.status === 'REFUNDED') {
    return {
      success: false,
      error: `Order '${orderId}' has already been refunded.`,
    };
  }

  // Apply state change
  const timestamp = now();
  orders[orderIdx] = {
    ...order,
    resolution_state: 'REFUND_PROCESSED',
    history: [
      ...order.history,
      {
        timestamp,
        action: 'PROCESS_REFUND',
        details: `Processed refund of $${refundAmount.toFixed(2)} to ${order.payment_method}.`,
      },
    ],
  };

  // Calculate VIP bonus credit
  const cust = customers.find((c) => c.customer_id === order.customer_id);
  const vipBonusPercent = cust?.tier === 'VIP' ? policiesData.refund_policy.vip_bonus_percent : 0;
  const vipBonus = (refundAmount * vipBonusPercent) / 100;

  return {
    success: true,
    order_id: orderId,
    refund_amount: refundAmount,
    vip_bonus_credit: vipBonus,
    total_credited: refundAmount + vipBonus,
    payment_method: order.payment_method,
    new_resolution_state: 'REFUND_PROCESSED',
    status: order.status,
    message: `Refund of $${refundAmount.toFixed(2)} successfully processed for order ${orderId}.${
      vipBonus > 0 ? ` VIP bonus credit of $${vipBonus.toFixed(2)} (10%) applied.` : ''
    }`,
  };
}

export function processReplacement(orderId: string, sku: string): ToolResult {
  const orderIdx = orders.findIndex((o) => o.order_id === orderId);
  if (orderIdx === -1) {
    return { success: false, error: `Order ID '${orderId}' not found.` };
  }
  const order = orders[orderIdx];

  // Prevent duplicate resolution
  if (order.resolution_state !== 'NONE') {
    return {
      success: false,
      error: `Order '${orderId}' already has resolution state '${order.resolution_state}'.`,
    };
  }

  // Verify policy
  const policyCheck = checkPolicy(orderId, 'REPLACEMENT');
  if (!policyCheck.success) {
    return policyCheck;
  }

  // Check inventory stock for requested SKU
  const inv = checkInventory(sku);
  if (!inv.success) {
    return inv;
  }

  // Mandatory honest failure check: If stock is 0, report clear failure
  const qty = typeof inv.quantity_available === 'number' ? inv.quantity_available : 0;
  const itemName = typeof inv.item_name === 'string' ? inv.item_name : sku;
  const restockDate = typeof inv.restock_date === 'string' ? inv.restock_date : null;

  if (!inv.in_stock || qty <= 0) {
    return {
      success: false,
      error: `Replacement failed: SKU '${sku}' (${itemName}) is out of stock (quantity: 0).`,
      out_of_stock_sku: sku,
      sku: sku,
      item_name: itemName,
      quantity_available: 0,
      restock_date: restockDate,
      suggestion: 'Consider offering a refund instead, or check alternative SKUs.',
    };
  }

  // Deduct 1 item from inventory
  const invIdx = inventory.findIndex((i) => i.sku === sku);
  if (invIdx !== -1) {
    inventory[invIdx] = {
      ...inventory[invIdx],
      quantity_in_stock: inventory[invIdx].quantity_in_stock - 1,
    };
  }

  // Update order state
  const timestamp = now();
  orders[orderIdx] = {
    ...order,
    resolution_state: 'REPLACEMENT_PROCESSED',
    history: [
      ...order.history,
      {
        timestamp,
        action: 'PROCESS_REPLACEMENT',
        details: `Replacement unit ordered for SKU '${sku}' (${itemName}). Shipping to ${order.shipping_address}.`,
      },
    ],
  };

  return {
    success: true,
    order_id: orderId,
    replacement_sku: sku,
    replacement_item: itemName,
    new_resolution_state: 'REPLACEMENT_PROCESSED',
    status: order.status,
    message: `Replacement for SKU '${sku}' (${itemName}) successfully created for order ${orderId}.`,
  };
}

export function cancelOrder(orderId: string): ToolResult {
  const orderIdx = orders.findIndex((o) => o.order_id === orderId);
  if (orderIdx === -1) {
    return { success: false, error: `Order ID '${orderId}' not found.` };
  }
  const order = orders[orderIdx];

  // Prevent duplicate cancellation
  if (order.status === 'CANCELLED' || order.resolution_state === 'CANCELLED') {
    return {
      success: false,
      error: `Order '${orderId}' has already been cancelled.`,
    };
  }

  // Check policy for cancellation
  const polCheck = checkPolicy(orderId, 'CANCEL');
  if (!polCheck.allowed) {
    return {
      success: false,
      error: `Cancellation blocked by policy: ${polCheck.error || polCheck.reason}`,
    };
  }

  // Execute cancellation
  const timestamp = now();
  orders[orderIdx] = {
    ...order,
    status: 'CANCELLED' as OrderStatus,
    resolution_state: 'CANCELLED',
    history: [
      ...order.history,
      {
        timestamp,
        action: 'CANCEL_ORDER',
        details: 'Order cancelled successfully prior to shipping.',
      },
    ],
  };

  return {
    success: true,
    order_id: orderId,
    refund_amount: order.total_amount,
    payment_method: order.payment_method,
    new_status: 'CANCELLED',
    new_resolution_state: 'CANCELLED',
    status: 'CANCELLED',
    message: `Order ${orderId} has been successfully cancelled. Refund of $${order.total_amount.toFixed(2)} processed.`,
  };
}

// ---- Verification Tool ----

export function verifyState(orderId: string): ToolResult {
  const order = orders.find((o) => o.order_id === orderId);
  if (!order) {
    return { success: false, error: `Order ID '${orderId}' not found.` };
  }
  const isResolved = ['REFUND_PROCESSED', 'REPLACEMENT_PROCESSED', 'CANCELLED', 'REFUNDED', 'REPLACED'].includes(
    order.resolution_state
  );

  return {
    success: true,
    order_id: orderId,
    status: order.status,
    current_status: order.status,
    resolution_state: order.resolution_state,
    is_resolved: isResolved,
    items: order.items.map((i) => ({ ...i })),
    total_amount: order.total_amount,
    last_action: order.history.length > 0 ? order.history[order.history.length - 1] : null,
    history_count: order.history.length,
    history: order.history.map((h) => ({ ...h })),
  };
}

// ---- Utility: Get complete data for state panel ----

export function getFullState(customerId: string, orderId: string) {
  const cust = customers.find((c) => c.customer_id === customerId);
  const order = orders.find((o) => o.order_id === orderId);
  const relevantSkus = order ? order.items.map((i) => i.sku) : [];
  const relevantInventory = inventory.filter((i) => relevantSkus.includes(i.sku));
  return { customer: cust ? { ...cust } : null, order: order ? { ...order } : null, inventory: relevantInventory.map((i) => ({ ...i })) };
}
